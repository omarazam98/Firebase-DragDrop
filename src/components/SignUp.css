#phoneNumber {
    background-color: orange
}

#email {
    background-color: red

}

#passwordOne {
    background-color: green;

}

#passwordTwo {
    background-color: blue

}

#phoneNumber p.ws-errormessage::after {
    content:"Please enter the phone number in the form (xxx) xxx-xxxx";
}

.error { border: 1px solid red; }

.form-control:valid~.invalid-feedback {display: none;}
