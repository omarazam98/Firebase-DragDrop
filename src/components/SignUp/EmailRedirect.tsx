import React, { Component } from 'react';
import { withRouter } from 'react-router-dom'

export function EmailRedirect(props) {
    const handleClick = (e) => props.history.push('/login');
    return (
        <div>
            Email has been sent to verify account. <br/>
            Already verified?
            <button onClick={handleClick}> Log In</button>
        </div>    );
}
export default withRouter(EmailRedirect);
