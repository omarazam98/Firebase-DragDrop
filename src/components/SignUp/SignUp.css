/*#phoneNumber {
    background-color: orange
}

#email {
    background-color: red

}

#passwordOne {
    background-color: green;

}

#passwordTwo {
    background-color: blue

}

*/
span {
    color: firebrick;
}
