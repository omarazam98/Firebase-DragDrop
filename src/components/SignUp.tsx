import React, { Component } from 'react';
import './SignUp.css';
import { fireauth, firestore} from '../firebase/firebase'

interface SignUpState {
    name: {
        value: string,
        touched: boolean,
        errors: string,
    };
    phoneNumber: {
        value: string,
        touched: boolean,
        errors: string,
    };
    class:{
        value: string,
        touched: boolean,
        errors: string,
    };
    email: {
        value: string,
        touched: boolean,
        errors: string,
    };
    passwordOne: {
        value: string,
        touched: boolean,
        errors: string,
    };
    passwordTwo: {
        value: string,
        touched: boolean,
        errors: string,
    };
}

const INITIAL_STATE: SignUpState = {
    name: {
        value: '',
        touched: false,
        errors: '',
    },
    phoneNumber: {
        value: '',
        touched: false,
        errors: '',
    },
    class: {
        value: '',
        touched: false,
        errors: '',
    },
    email: {
        value: '',
        touched: false,
        errors: '',
    },
    passwordOne: {
        value: '',
        touched: false,
        errors: '',
    },
    passwordTwo: {
        value: '',
        touched: false,
        errors: '',
    },
};

class SignUp extends Component<any, SignUpState> {
    constructor(props) {
        super(props);
        this.state = INITIAL_STATE;
    }

    onChange = e => {
        e.persist()

        if (Object.keys(this.state).includes(e.target.name)) {
            this.setState({
                [e.target.name]: {
                    ...this.state[e.target.name],
                    value: e.target.value
                }
            } as Pick<SignUpState, keyof SignUpState>, () => {
                this.validateInputs(e);
            });

        };
    };
    validateInputs = e => {
        if (e.target.value === '') {
            this.setState({
                [e.target.name]: {
                    ...this.state[e.target.name],
                    errors: 'field cannot be empty'
                }
            } as Pick<SignUpState, keyof SignUpState>);
        }else if (e.target.name === 'passwordTwo' && this.state.passwordOne.value !== this.state.passwordTwo.value) {
            this.setState({
                passwordTwo: {
                    ...this.state.passwordTwo,
                    errors: 'passwords must match'
                }
            } as Pick<SignUpState, keyof SignUpState>);
        } else if (!e.target.checkValidity()) {
            this.setState({
                [e.target.name]: {
                    ...this.state[e.target.name],
                    errors: 'invalid format'
                }
            } as Pick<SignUpState, keyof SignUpState>);
        } else {
            this.setState({
                [e.target.name]: {
                    ...this.state[e.target.name],
                    errors: ''
                }
            } as Pick<SignUpState, keyof SignUpState>);
        };
    };
    submit = e => {
        fireauth.createUserWithEmailAndPassword(this.state.email.value, this.state.passwordOne.value).then((userCredential) => {
            const user = userCredential.user;
            if (user){
                user.updateProfile({ displayName: this.state.name.value }).then(()=>{
                    console.log(user.uid);
                    console.log(user.emailVerified);
                    firestore.collection('users').doc(user.uid).set({email: user.email, phoneNumber: this.state.phoneNumber.value, name: user.displayName, class: this.state.class.value});
                });
            }
        }).catch(function(error) {
            var errorCode = error.code;
            var errorMessage = error.message;
            if (errorCode == 'auth/weak-password') {
                alert('The password is too weak.');
            } else {
                alert(errorMessage);
            }
        });
        e.preventDefault();
    }

    hasErrors = () => {
        for(const key in this.state){
            if (this.state[key].errors || this.state[key].value === ''){
                return true;
            }
        }
        return false;
    };

    handleBlur = e => {
        if (Object.keys(this.state).includes(e.target.name)) {
            this.setState({[e.target.name]: { ...this.state[e.target.name], touched: true}} as Pick<SignUpState, keyof SignUpState>);
        };
    };

    render(){
        const disableButton = this.hasErrors();
        return(
            <div>
                <h1> Sign Up Page </h1>
                <form onSubmit={this.submit}>
                    <label htmlFor="name">Name: </label> <br/>
                    <input name="name" id="name" onBlur = {this.handleBlur} value={this.state.name.value} onChange={this.onChange} type="text" placeholder="Full Name"/> {this.state.name.touched && <span> {this.state.name.errors} </span> } <br/>

                    <input type="radio" name="class" value="senior" checked={this.state.class.value === "senior"} onChange={this.onChange} />Senior
                    <input type="radio" name="class" value="student" checked={this.state.class.value === "student"} onChange={this.onChange} />Student <br/>

                    <label htmlFor="email">Email: </label> <br/>
                    <input name="email" id="email" onBlur = {this.handleBlur} value={this.state.email.value} onChange={this.onChange} type="email" placeholder="Email Address"/>  {this.state.email.touched && <span> {this.state.email.errors} </span> }<br/>
                    <label htmlFor="phoneNumber">Phone Number: </label><br/>
                    <input name="phoneNumber" id='phoneNumber' onBlur={this.handleBlur} onChange={this.onChange} type="tel" placeholder="Phone Number" pattern="[0-9]{3}-[0-9]{3}-[0-9]{4}"/>{this.state.phoneNumber.touched && <span> {this.state.phoneNumber.errors} </span> }<br/>
                    <label htmlFor="passwordOne">Password: </label><br/>
                    <input name="passwordOne" id="passwordOne" onBlur={this.handleBlur} value={this.state.passwordOne.value} onChange={this.onChange} type="password" placeholder="Password" />{this.state.passwordOne.touched && <span> {this.state.passwordOne.errors} </span> } <br/>
                    <label htmlFor="passwordTwo">Confirm Password: </label> <br/>
                    <input name="passwordTwo" id="passwordTwo"onBlur={this.handleBlur} value={this.state.passwordTwo.value} onChange={this.onChange} type="password" placeholder="Confirm Password" />{ this.state.passwordTwo.touched && <span> {this.state.passwordTwo.errors} </span> } <br/>
                    <input disabled={disableButton} type="submit" value={'Sign Up'}></input>
        </form>
            </div>
        );
    };
}
export default SignUp;
