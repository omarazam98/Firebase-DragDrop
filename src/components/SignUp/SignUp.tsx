import React, { Component } from 'react';
import '../../styles/SignUp.css';
import { withAPI } from '@winwin/api-firebase';
import { withRouter } from 'react-router-dom';

interface FormField {
    value: string,
    touched: boolean,
    errors: string,
}
interface SignUpState {
    name: FormField;
    phoneNumber: FormField;
    class: FormField;
    email: FormField;
    passwordOne: FormField;
    passwordTwo: FormField;
}

const INITIAL_STATE: SignUpState = {
    name: {
        value: '',
        touched: false,
        errors: '',
    },
    phoneNumber: {
        value: '',
        touched: false,
        errors: '',
    },
    class: {
        value: '',
        touched: false,
        errors: '',
    },
    email: {
        value: '',
        touched: false,
        errors: '',
    },
    passwordOne: {
        value: '',
        touched: false,
        errors: '',
    },
    passwordTwo: {
        value: '',
        touched: false,
        errors: '',
    },
};

export class SignUp extends Component<any, SignUpState> {
    constructor(props) {
        super(props);
        this.state = INITIAL_STATE;
        this.handleChange = this.handleChange.bind(this);
        this.validateInputs = this.validateInputs.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleBlur = this.handleBlur.bind(this);
        this.hasErrors = this.hasErrors.bind(this);
        this.updateField = this.updateField.bind(this);
    }

    handleChange(e) {
        e.persist()
        if (Object.keys(this.state).includes(e.target.name)) {
            this.updateField(e.target.name, e.target.value, e.target.checkValidity())
        }
    };

    updateField(name, value, valid) {
        this.setState(function (prevState, props) {
            return {
                [name]: {
                    ...prevState[name],
                    value
                }
            } as Pick<SignUpState, keyof SignUpState>
        }, () => {
            this.validateInputs(name, value, valid)
        });
    }

    validateInputs(name, value, valid) {
        if (value === '') {
            this.setState(function (prevState, props) {
                return {
                    [name]: {
                        ...prevState[name],
                        errors: 'field cannot be empty'
                    }
                } as Pick<SignUpState, keyof SignUpState>
            });
        } else if (name === 'passwordTwo' && this.state.passwordOne.value !== this.state.passwordTwo.value) {
            this.setState(function (prevState, props) {
                return {
                    passwordTwo: {
                        ...prevState.passwordTwo,
                        errors: 'passwords must match'
                    }
                } as Pick<SignUpState, keyof SignUpState>
            });
        } else if (!valid) {
            this.setState(function (prevState, props) {
                return {
                    [name]: {
                        ...prevState[name],
                        errors: 'invalid format'
                    }
                } as Pick<SignUpState, keyof SignUpState>
            });
        } else {
            this.setState(function (prevState, props) {
                return {
                    [name]: {
                        ...prevState[name],
                        errors: ''
                    }
                } as Pick<SignUpState, keyof SignUpState>
            });
        };
    };

    handleSubmit(e) {
        e.preventDefault();
        return this.props.api.auth.signup.createUserWithEmailAndPassword(this.state.email.value, this.state.passwordOne.value)
                .then((userCredential) => {
                    const user = userCredential.user;
                    if (user) {
                        user.sendEmailVerification();
                        this.props.history.push('/email'); //redirects to a new page saying that an email has been sent, and the account needs to be verified
                        user.updateProfile({displayName: this.state.name.value}).then(() => { //after creating the user in firebase auth, this also starts a user profile in the DB that will add the extra info (name, phonenumber, student or senior)
                            this.props.api.data.users.create({
                                email: user.email,
                                phoneNumber: this.state.phoneNumber.value,
                                name: user.displayName,
                                class: this.state.class.value
                            }, user.uid);
                        });
                    }
                }).catch(function (error) { //these error messages are coming from firestore and include anything we aren't already catching
                    var errorCode = error.code;
                    var errorMessage = error.message;
                    if (errorCode == 'auth/weak-password') {
                        alert('The password is too weak.');
                    } else {
                        alert(errorMessage);
                    }
                });
    }

    hasErrors(): boolean {
        for (const key in this.state) {
            if (this.state[key].errors || this.state[key].value === '') {
                return true;
            }
        }
        return false;
    };

    /*
        handleBlur is essentially an onUnFocus method so that errors aren't rendered right away, just after the user goes away from the input box
     */
    handleBlur(e) { //
        e.persist();
        if (Object.keys(this.state).includes(e.target.name)) {
            this.setState(function (prevState, props) {
                return {
                    [e.target.name]: {
                        ...prevState[e.target.name],
                        touched: true
                    }
                } as Pick<SignUpState, keyof SignUpState>
            });
        }

    }

    render() {
        console.log(this.state);
        const disableButton = this.hasErrors();
        return (
            <div>
                <h1> Sign Up Page </h1>
                <form onSubmit={this.handleSubmit}>
                    <label htmlFor="name">Name: </label> <br/>
                    <input name="name" id="name" onBlur={this.handleBlur} value={this.state.name.value}
                           onChange={this.handleChange} type="text" placeholder="Full Name"/> {this.state.name.touched &&
                <span> {this.state.name.errors} </span>} <br/>

                    <input type="radio" name="class" value="senior" checked={this.state.class.value === "senior"}
                           onChange={this.handleChange}/>Senior
                    <input type="radio" name="class" value="student" checked={this.state.class.value === "student"}
                           onChange={this.handleChange}/>Student <br/>

                    <label htmlFor="email">Email: </label> <br/>
                    <input name="email" id="email" onBlur={this.handleBlur} value={this.state.email.value}
                           onChange={this.handleChange} type="email"
                           placeholder="Email Address"/> {this.state.email.touched &&
                <span> {this.state.email.errors} </span>}<br/>

                    <label htmlFor="phoneNumber">Phone Number: </label><br/>
                    <input name="phoneNumber" id='phoneNumber' onBlur={this.handleBlur} onChange={this.handleChange}
                           type="tel" placeholder="Phone Number"
                           pattern="[0-9]{3}-[0-9]{3}-[0-9]{4}"/>{this.state.phoneNumber.touched &&
                <span> {this.state.phoneNumber.errors} </span>}<br/>

                    <label htmlFor="passwordOne">Password: </label><br/>
                    <input name="passwordOne" id="passwordOne" onBlur={this.handleBlur}
                           value={this.state.passwordOne.value} onChange={this.handleChange} type="password"
                           placeholder="Password"/>{this.state.passwordOne.touched &&
                <span> {this.state.passwordOne.errors} </span>} <br/>

                    <label htmlFor="passwordTwo">Confirm Password: </label> <br/>
                    <input name="passwordTwo" id="passwordTwo" onBlur={this.handleBlur}
                           value={this.state.passwordTwo.value} onChange={this.handleChange} type="password"
                           placeholder="Confirm Password"/>{this.state.passwordTwo.touched &&
                <span> {this.state.passwordTwo.errors} </span>} <br/>

                    <input disabled={disableButton} id={'submitButton'} type="submit"
                           value={'Sign Up'}></input>
                </form>
            </div>
        );
    };
}

export default withAPI(withRouter(SignUp))
