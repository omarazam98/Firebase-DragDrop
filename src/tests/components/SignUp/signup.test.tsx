import React from 'react';
import ReactDOM from 'react-dom';
import { SignUp } from '../../../components/SignUp/SignUp';
import {shallow} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16';

let wrapper;

const fakeAPI = {
        auth: {
            signup: {
                createUserWithEmailAndPassword: () => {
                    return new Promise((resolve, reject) => {
                        const userCredential = {
                            user: {
                                name: 'sample user',
                                sendEmailVerification: jest.fn(),
                            }
                        }
                        resolve(userCredential);
                    });
                }
            }
        }
};

const mockState = {
    name: {
        value: 'Julia St-Jean',
        touched: true,
        errors: '',
    },
    phoneNumber: {
        value: '613-123-4567',
        touched: true,
        errors: '',
    },
    class: {
        value: 'student',
        touched: true,
        errors: '',
    },
    email: {
        value: 'sample@email.com',
        touched: true,
        errors: '',
    },
    passwordOne: {
        value: 'aaaaaa',
        touched: true,
        errors: '',
    },
    passwordTwo: {
        value: 'aaaaaa',
        touched: true,
        errors: '',
    },
};

beforeEach(() => {
    wrapper = shallow(
        <SignUp/>);
});

test('has a valid snapshot', () => {
    expect(wrapper).toMatchSnapshot();
});

test('redirects to login after submit', () => {
    const fakeEvent = { preventDefault: () => {} };
    const historyMock = { push: jest.fn() };
    const firebaseMock = { }
    const historyWrapper = shallow(<SignUp history={historyMock} api={fakeAPI}/>)
    historyWrapper.setState(mockState);
    historyWrapper.instance().handleSubmit(fakeEvent).then(()=>{
        expect(historyMock.push.mock.calls[0]).toEqual([ '/email' ]);
    })
});

