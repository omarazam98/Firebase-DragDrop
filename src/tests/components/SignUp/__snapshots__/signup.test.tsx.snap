// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Email Redirect has a valid snapshot 1`] = `
<div>
  Email has been sent to verify account. 
  <br />
  Already verified?
  <button
    id="redirect"
    onClick={[Function]}
  >
     Log In
  </button>
   
  <br />
  Didn't receive the email?
  <button
    id="resend"
    onClick={[Function]}
  >
     Send Again
  </button>
</div>
`;

exports[`has a valid snapshot 1`] = `
<div>
  <h1>
     Sign Up 
  </h1>
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="name"
    >
      Name: 
    </label>
     
    <br />
    <input
      id="name"
      name="name"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Full Name"
      type="text"
      value=""
    />
     
     
    <br />
    <input
      checked={false}
      name="class"
      onChange={[Function]}
      onFocus={[Function]}
      type="radio"
      value="senior"
    />
    Senior
    <input
      checked={false}
      name="class"
      onChange={[Function]}
      onFocus={[Function]}
      type="radio"
      value="student"
    />
    Student
     
    <br />
    <label
      htmlFor="email"
    >
      Email: 
    </label>
     
    <br />
    <input
      id="email"
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Email Address"
      type="email"
      value=""
    />
    <br />
    <label
      htmlFor="phoneNumber"
    >
      Phone Number: 
    </label>
    <br />
    <input
      id="phoneNumber"
      name="phoneNumber"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      pattern="[0-9]{3}-[0-9]{3}-[0-9]{4}"
      placeholder="Phone Number"
      type="tel"
    />
     
    <br />
    <label
      htmlFor="passwordOne"
    >
      Password: 
    </label>
    <br />
    <input
      id="passwordOne"
      name="passwordOne"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
     
    <br />
    <label
      htmlFor="passwordTwo"
    >
      Confirm Password: 
    </label>
     
    <br />
    <input
      id="passwordTwo"
      name="passwordTwo"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Confirm Password"
      type="password"
      value=""
    />
     
    <br />
    <input
      id="submitButton"
      type="submit"
      value="Sign Up"
    />
     
    <span />
  </form>
</div>
`;
